prow/github/client.go

```diff
@@ -185,18 +190,36 @@ func (c *Client) Throttle(hourlyTokens, burst int) {
 	c.throttle.throttle = throttle
 }
 
+// RoundTrip is an implementation for net/http RoundTripper interface.
+// Its being used to refresh the token from a token source.
+func (ct *cacherTransport) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	_, err = ct.Base.Source.Token()
+	if err != nil {
+		return nil, err
+	}
+	// ref: https://godoc.org/net/http#Transport.RoundTrip
+	resp, err = ct.Base.RoundTrip(req)
+	if err != nil {
+		return nil, err
+	}
+	_, err = ct.Base.Source.Token()
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
 // NewClient creates a new fully operational GitHub client.
 // 'getToken' is a generator for the GitHub access token to use.
 // 'bases' is a variadic slice of endpoints to use in order of preference.
 //   An endpoint is used when all preceding endpoints have returned a conn err.
 //   This should be used when using the ghproxy GitHub proxy cache to allow
 //   this client to bypass the cache if it is temporarily unavailable.
-func NewClient(getToken func() []byte, bases ...string) *Client {
+func NewClient(getToken func() []byte, tokenSource oauth2.TokenSource, bases ...string) *Client {
```

Comment:
This signature is a little odd, we shouldn't make it possible for someone to pass a separate token for the graphql bits than the other one -- maybe just take the SecretAgent and use it's methods inside of the constructor

Reply:
The SecretAgent must implement the TokenSource interface in order for the roundtrip to be possible. If I catch your point correctly, we can't use the SecretAgent's methods to dynamically reload the token in the graphql client.

Reply:
There's no reason I can't pass in something entirely unrelated to `getToken` as to `tokenSource`. Make this func take the `secretAgent` so we know we are using the same token source for both